src/Aspire.Hosting/ApplicationModel/InteractionService.cs

```diff
@@ -0,0 +1,584 @@
+    /// <returns>
+    /// An <see cref="InteractionResult{T}"/> containing <c>true</c> if the user accepted, <c>false</c> otherwise.
+    /// </returns>
+    public async Task<InteractionResult<bool>> PromptMessageBarAsync(string title, string message, MessageBoxInteractionOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        options ??= MessageBoxInteractionOptions.CreateDefault();
+
+        var newState = new Interaction(title, message, options, new Interaction.MessageBarInteractionInfo(intent: options.Intent ?? MessageIntent.None), cancellationToken);
+        AddInteractionUpdate(newState);
+
+        using var _ = cancellationToken.Register(OnInteractionCancellation, state: newState);
+
+        var completion = await newState.CompletionTcs.Task.ConfigureAwait(false);
+        return completion.Canceled
+            ? InteractionResultFactory.Cancel<bool>()
+            : InteractionResultFactory.Ok((bool)completion.State!);
+    }
+
+    // For testing.
+    internal List<Interaction> GetCurrentInteractions()
+    {
+        lock (_onInteractionUpdatedLock)
+        {
+            return _interactionCollection.ToList();
+        }
+    }
+
+    private void OnInteractionCancellation(object? newState)
+    {
+        var interactionState = (Interaction)newState!;
+
+        interactionState.State = Interaction.InteractionState.Complete;
+        interactionState.CompletionTcs.TrySetResult(new InteractionCompletionState { Canceled = true });
+        AddInteractionUpdate(interactionState);
+    }
+
+    private void AddInteractionUpdate(Interaction interactionUpdate)
+    {
+        lock (_onInteractionUpdatedLock)
+        {
+            var updateEvent = false;
+
+            if (interactionUpdate.State == Interaction.InteractionState.Complete)
+            {
+                Debug.Assert(
+                    interactionUpdate.CompletionTcs.Task.IsCompleted,
+                    "TaskCompletionSource should be completed when interaction is done.");
+
+                // Only update event if interaction was previously registered and not already removed.
+                updateEvent = _interactionCollection.Remove(interactionUpdate.InteractionId);
+            }
+            else
+            {
+                if (_interactionCollection.Contains(interactionUpdate.InteractionId))
+                {
+                    // Should never happen, but throw descriptive exception if it does.
+                    throw new InvalidOperationException($"An interaction with ID {interactionUpdate.InteractionId} already exists. Interaction IDs must be unique.");
+                }
+
+                _interactionCollection.Add(interactionUpdate);
+                updateEvent = true;
+            }
+
+            if (updateEvent)
+            {
+                OnInteractionUpdated?.Invoke(interactionUpdate);
```

Comment:
Do we want to call this under the lock? What code executes here? Is it user code?

Reply:
I see, that's fine then